bootloaders/uart/uart_boot.c

```diff
@@ -0,0 +1,152 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include "pico/stdlib.h"
+#include "hardware/uart.h"
+#include "pico/bootrom.h"
+#include "boot/picobin.h"
+#include "hardware/flash.h"
+
+// UART defines for uart boot
+#define UART_ID uart1
+#define BAUD_RATE 1000000
+
+// Use pins 4 and 5 for uart boot
+#define UART_TX_PIN 4
+#define UART_RX_PIN 5
+
+// Use pin 3 for the RUN pin on the other chip
+#define RUN_PIN 3
+
+
+void reset_chip() {
+    // Toggle run pin
+    gpio_put(RUN_PIN, false);
+    sleep_ms(1);
+    gpio_put(RUN_PIN, true);
+}
+
+
+void uart_boot() {
+    uint knocks = 0;
+    while (true) {
+        // Send the knock sequence
+        uart_putc_raw(UART_ID, 0x56);
+        uart_putc_raw(UART_ID, 0xff);
+        uart_putc_raw(UART_ID, 0x8b);
+        uart_putc_raw(UART_ID, 0xe4);
+        uart_putc_raw(UART_ID, 'n');
+
+        if (uart_is_readable_within_us(UART_ID, 1000)) {
+            char in = uart_getc(UART_ID);
+            printf("%c\n", in);
+            break;
+        } else {
+            if (knocks > 10) {
+                printf("No response - resetting\n");    
+                reset_chip();
+                return;
+            }
+            printf("No response - knocking again\n");
+            knocks++;
+        }
+    }
+
+    printf("Boot starting\n");
+
+    // Get partition location in flash
+    const int buf_words = (16 * 4) + 1;   // maximum of 16 partitions, each with maximum of 4 words returned, plus 1
+    uint32_t* buffer = malloc(buf_words * 4);
+
+    int ret = rom_get_partition_table_info(buffer, buf_words, PT_INFO_PARTITION_LOCATION_AND_FLAGS | PT_INFO_SINGLE_PARTITION | (0 << 24));
+    assert(buffer[0] == (PT_INFO_PARTITION_LOCATION_AND_FLAGS | PT_INFO_SINGLE_PARTITION));
+    assert(ret == 3);
```

Comment:
What does the magic value 3 indicate? :wink:

Reply:
This one isn't actually a magic value - it's just the number of words written in the buffer (which should be 3 here - 1 word for the returned flags, and 2 words for the partition_location_and_flags)

Reply:
I guess it might be useful if https://github.com/raspberrypi/pico-sdk/blob/develop/src/rp2_common/pico_bootrom/include/pico/bootrom.h#L628 documented the return value.

Reply:
It does mention it https://github.com/raspberrypi/pico-sdk/blob/develop/src/rp2_common/pico_bootrom/include/pico/bootrom.h#L634

Reply:
Oh, I missed that :laughing:  My eyes just jumped straight to the bottom of the doxygen description looking for the `\return ...` line :stuck_out_tongue_winking_eye: